source/beamline/BioXAS/BioXASCarbonFilterFarm.cpp

```diff
@@ -0,0 +1,395 @@
+		upstreamFilterMap_.insert(window, filterThickness);
+
+	} else if (actuator == Actuator::Downstream) {
+		downstreamFilterMap_.insert(window, filterThickness);
+	}
+}
+
+void BioXASCarbonFilterFarmControl::setMoveInProgress(bool isMoving)
+{
+	if (moveInProgress_ != isMoving) {
+		moveInProgress_ = isMoving;
+		emit movingChanged(moveInProgress_);
+
+		if (moveInProgress_)
+			emit moveStarted();
+	}
+}
+
+void BioXASCarbonFilterFarmControl::onMoveStarted()
+{
+	setMoveInProgress(true);
+}
+
+void BioXASCarbonFilterFarmControl::onMoveCancelled(QObject *action)
+{
+	moveCleanup(action);
+	emit moveFailed(AMControl::WasStoppedFailure);
+}
+
+void BioXASCarbonFilterFarmControl::onMoveFailed(QObject *action)
+{
+	moveCleanup(action);
+	emit moveFailed(AMControl::OtherFailure);
+}
+
+void BioXASCarbonFilterFarmControl::onMoveSucceeded(QObject *action)
+{
+	moveCleanup(action);
+	emit moveSucceeded();
+}
+
+void BioXASCarbonFilterFarmControl::updateFilter()
+{
+	Filter::Thickness newFilter = Filter::Invalid;
+
+	if (upstreamActuator_ && upstreamActuator_->isConnected() && downstreamActuator_ && downstreamActuator_->isConnected()) {
+		BioXASCarbonFilterFarmActuatorControl::Window::Selection upstreamWindow = upstreamActuator_->window(upstreamActuator_->value());
+		BioXASCarbonFilterFarmActuatorControl::Window::Selection downstreamWindow = downstreamActuator_->window(downstreamActuator_->value());
+
+		newFilter = calculateTotalFilterFromWindows(upstreamWindow, downstreamWindow);
+	}
+
+	setFilter(newFilter);
+}
+
+BioXASCarbonFilterFarmControl::Filter::Thickness BioXASCarbonFilterFarmControl::calculateTotalFilterFromWindows(BioXASCarbonFilterFarmActuatorControl::Window::Selection upstreamWindow, BioXASCarbonFilterFarmActuatorControl::Window::Selection downstreamWindow)
+{
+	Filter::Thickness upstreamThickness = Filter::Invalid;
+	Filter::Thickness downstreamThickness = Filter::Invalid;
+
+	if (upstreamActuator_ && upstreamActuator_->validWindow(upstreamWindow) && upstreamFilterMap_.contains(upstreamWindow)) {
+		upstreamThickness = filterAtWindow(Actuator::Upstream, upstreamWindow);
+	}
+
+	if (downstreamActuator_ && downstreamActuator_->validWindow(downstreamWindow) && downstreamFilterMap_.contains(downstreamWindow)) {
+		downstreamThickness = filterAtWindow(Actuator::Downstream, downstreamWindow);
+	}
+
+	return calculateTotalFilter(upstreamThickness, downstreamThickness);
+}
+
+BioXASCarbonFilterFarmControl::Filter::Thickness BioXASCarbonFilterFarmControl::calculateTotalFilter(Filter::Thickness upstreamFilter, Filter::Thickness downstreamFilter)
+{
+	return doubleToFilter(filterToDouble(upstreamFilter) + filterToDouble(downstreamFilter));
+}
+
+void BioXASCarbonFilterFarmControl::moveCleanup(QObject *action)
+{
+	if (action) {
+		disconnect( action, 0, cancelledMapper_, 0 );
+		disconnect( action, 0, failedMapper_, 0 );
+		disconnect( action, 0, succeededMapper_, 0 );
+
+		cancelledMapper_->removeMappings(action);
```

Comment:
Since there is only one action being used at any given instance, the signal mapper may be a bit overkill since you probably could replace it with a individual connect signals and just keep the action as a member variable.  It still works though and it does easily give you access to the pointer of the action once it's done - which is the important thing for the memory management aspect that you've already handled.